common/base/include/claragenomics/utils/device_preallocated_allocator.cuh

```diff
@@ -78,14 +82,16 @@ public:
     /// Memory allocation is aligned by 256 bytes
     /// \param ptr on return pointer to allocated memory, nullptr if allocation was not successful
     /// \param bytes_needed
-    /// \param associated_stream on deallocation this block will be free only once all work in this stream has finished
+    /// \param associated_streams on deallocation this block will be freed only once all previosly scheduled work in these streams has finished, using default stream as default
     /// \return cudaSuccess if allocation was successful, cudaErrorMemoryAllocation otherwise
     cudaError_t DeviceAllocate(void** ptr,
                                size_t bytes_needed,
-                               cudaStream_t associated_stream = 0)
+                               const std::vector<cudaStream_t>& associated_streams = {{0}})
     {
+        assert(!associated_streams.empty());
+
         std::lock_guard<std::mutex> mutex_lock_guard(memory_operation_mutex_);
-        return get_free_block(ptr, bytes_needed, associated_stream);
+        return get_free_block(ptr, bytes_needed, associated_streams);
```

Comment:
Optional/off-topic: Now reading the code, I stumbled over `get_free_block`. How about renaming the function to `allocate_free_block`?

Reply:
What about `allocate_memory_block()` and `free_memory_block`.
For me it is more confusing that we have `get_free_block()` and `free_block()`, when one `free` is adjective and the verb

Reply:
Sounds good to me.